storage-fetch/src/peer_resolver.rs

```diff
@@ -0,0 +1,111 @@
+use std::time::Duration;
+
+use anyhow::bail;
+use libp2p::{
+    futures::StreamExt,
+    noise,
+    request_response::{self, Message, ProtocolSupport},
+    swarm::SwarmEvent,
+    tcp, yamux, Multiaddr, PeerId, StreamProtocol, Swarm, SwarmBuilder,
+};
+use primitives::p2p::{PeerIdRequest, PeerInfoResponse, BOOTSTRAP_REQUEST_RESPONSE_PROTOCOL};
+use tracing::{info, warn};
+
+pub async fn find_multiaddr_storage_provider(
+    bootstrap_address: Multiaddr,
+    bootstrap_peer_id: PeerId,
+    sp_peer_id: PeerId,
+) -> Result<Vec<Multiaddr>, anyhow::Error> {
+    let swarm = create_swarm()?;
+
+    let peer_info = run_discover(swarm, bootstrap_address, &bootstrap_peer_id, sp_peer_id)
+        .await
+        .unwrap();
+
+    match peer_info {
+        PeerInfoResponse::NotFound(peer) => {
+            warn!(?peer, "peer is not registered");
+            bail!("{:?} is not registered", peer)
```

Comment:
The `bail` in this precise spot is slightly more confusing because it adds another layer of indirection in the already final return. Normally, you'd use it for early returns instead
```suggestion
            Err(anyhow!("{:?} is not registered", peer))
```
In contrast, the bails in lines 91 and 99 are better usage examples

Reply:
Nice one. This was the thing I was asking myself when writing this